### PR TITLE
Add trailing newline check for `rustc --print` in `run-make-support`

### PR DESCRIPTION
**Pull request by rejyr**
From [this comment](https://github.com/rust-lang/rust/pull/127877#issuecomment-2234363580).

Checks that `rustc --print` stdout ends with a newline. 
Adds `CompletedProcess` check closures to `Command`. 
Unfortunately, this means that `#[derive(Debug)]` macro for `Command` no longer works.  

Part of #121876.

r? @jieyouxu

### CODE REVIEW

**Review by jieyouxu — CHANGES_REQUESTED** [review]
I don't really like the `Vec<(String, Box<dyn Fn(&CompletedProcess)>)>` related extra complexity here. Why is the

```rs
assert!(out.stdout_utf8().ends_with('\n'));
```

check not sufficient?

**Comment by Rejyr** [comment]
> I don't really like the `Vec<(String, Box<dyn Fn(&CompletedProcess)>)>` related extra complexity here. Why is the
> 
> ```rust
> assert!(out.stdout_utf8().ends_with('\n'));
> ```
> 
> check not sufficient?

This checks that every `rustc` invocation with `--print` is checked with a newline, instead of manually for every invocation.

The `Vec<(String, Box<dyn Fn(&CompletedProcess)>)>` is not great, it could be implemented with
```rust
#[derive(Debug)] // this works now
pub struct Command {
    cmd: StdCommand,
    stdin: Option<Box<[u8]>>,
    assert_stdout_ends_with_newline: bool,
    drop_bomb: DropBomb,
}
```
and 
```rust
// Command::command_output
if self.assert_stdout_ends_with_newline {
    completed_process.assert_stdout_ends_with_newline();
}
```

File: src/tools/run-make-support/src/command.rs [hunk 1]
```diff
@@ -24,18 +24,50 @@ use build_helper::drop_bomb::DropBomb;
 /// [`run`]: Self::run
 /// [`run_fail`]: Self::run_fail
 /// [`run_unchecked`]: Self::run_unchecked
-#[derive(Debug)]
 pub struct Command {
     cmd: StdCommand,
     stdin: Option<Box<[u8]>>,
+    completed_process_checks: Vec<(String, Box<dyn Fn(&CompletedProcess)>)>,
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/command.rs:30]: > ```rust
> #[derive(Debug)] // this works now
> pub struct Command {
>     cmd: StdCommand,
>     stdin: Option<Box<[u8]>>,
>     assert_stdout_ends_with_newline: bool,
>     drop_bomb: DropBomb,
> }
> ```
> 
> and
> 
> ```rust
> // Command::command_output
> if self.assert_stdout_ends_with_newline {
>     completed_process.assert_stdout_ends_with_newline();
> }
> ```

Let's use a simple bool for now, if there are more check-hooks that we wish to register upon construction, then we can always refactor into the hook form when it comes to it.

@rustbot author

**Comment by JohnCSimon** [comment]
@Rejyr 
ping from triage - can you post your status on this PR? This PR has not received an update in a few months.


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

Or if you're not going to continue, please close it. Thank you!

**Comment by Rejyr** [comment]
I think checking `rustc --print`  would be better accomplished with a dedicated test for all options instead of a hook.

**Pull request closed without merging**